src/commands.rs

```diff
@@ -2028,30 +2029,91 @@ fn cmd_commit(ui: &mut Ui, command: &CommandHelper, args: &CommitArgs) -> Result
     Ok(())
 }
 
+fn resolve_multiple_rewriteable_revsets(
+    revision_args: &[RevisionArg],
+    workspace_command: &WorkspaceCommandHelper,
+) -> Result<Vec<Commit>, CommandError> {
+    let mut acc = Vec::new();
+    for revset in revision_args {
+        let revisions = workspace_command.resolve_revset(revset)?;
+        workspace_command.check_non_empty(&revisions)?;
+        for commit in &revisions {
+            workspace_command.check_rewriteable(commit)?;
+        }
+        acc.extend(revisions);
+    }
+    Ok(acc)
+}
+
 fn cmd_duplicate(
     ui: &mut Ui,
     command: &CommandHelper,
     args: &DuplicateArgs,
 ) -> Result<(), CommandError> {
     let mut workspace_command = command.workspace_helper(ui)?;
-    let predecessor = workspace_command.resolve_single_rev(&args.revision)?;
-    workspace_command.check_rewriteable(&predecessor)?;
+    let mut to_duplicate: IndexSet<Commit> =
+        resolve_multiple_rewriteable_revsets(&args.revisions, &workspace_command)?
+            .into_iter()
+            .collect();
+    let mut duplicated_old_to_new: IndexMap<Commit, Commit> = IndexMap::new();
+
     let mut tx = workspace_command
-        .start_transaction(&format!("duplicate commit {}", predecessor.id().hex()));
+        .start_transaction(&format!("duplicating {} commit(s)", to_duplicate.len()));
     let mut_repo = tx.mut_repo();
-    let new_commit = mut_repo
-        .rewrite_commit(command.settings(), &predecessor)
-        .generate_new_change_id()
-        .write()?;
-    ui.write("Created: ")?;
-    write_commit_summary(
-        ui.stdout_formatter().as_mut(),
-        mut_repo.as_repo_ref(),
-        &workspace_command.workspace_id(),
-        &new_commit,
-        command.settings(),
-    )?;
-    ui.write("\n")?;
+    while !to_duplicate.is_empty() {
+        let mut original_commit = None;
+        // Find a commit whose parents we either don't plan to duplicate or have already
+        // duplicated.
+        // This reimplements the heads() revset function, but we will optimize it
+        // shortly.
+        for commit in to_duplicate.iter() {
+            if !commit
+                .parents()
+                .iter()
+                .any(|parent| to_duplicate.contains(parent))
+            {
+                original_commit = Some(commit.clone());
+                break;
+            }
+        }
```

Comment:
You should be able to call `repo.index().topo_order()` to get a topological order. I don't think that de-duplicates commits, so you'll still have to take care of that here (skipping commits already in `duplicated_old_to_new`).

Reply:
Done. This shortened things a bit.